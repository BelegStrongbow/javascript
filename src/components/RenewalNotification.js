import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import colors from "yoast-components/style-guide/colors.json";
import { FormattedMessage, FormattedDate, defineMessages, injectIntl, intlShape } from "react-intl";
import Link from "./Link";
import { CloseButtonTopRight } from "./Button";
import NewTabMessage from "../components/NewTabMessage";
import renewalNotification from "../images/renewalNotification.svg";

const messages = defineMessages( {
	header: {
		id: "renewal.notification.header",
		defaultMessage: "One or more plugins are about to expire!",
	},
	description: {
		id: "renewal.notification.description",
		defaultMessage: "When they do, you will no longer receive (security) updates or support! The first one expires on {expiryDate}.",
	},
	close: {
		id: "renewal.notification.close",
		defaultMessage: "close",
	},
	linkPlaceholder: {
		id: "renewal.notification.link.placeholder",
		defaultMessage: "{renewalLink}",
	},
	linkMessage: {
		id: "renewal.notification.link.message",
		defaultMessage: "Renew your subscription(s) on this page.",
	},
} );

const MessageContainer = styled.div`
	background-color: ${ colors.$color_white };
	box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
	padding: 8px 32px;
	margin-bottom: 32px;
	position: relative;
	display:flex;

	button {
		position: absolute;
		right: 16px;
		top: 16px;
	}
`;

const MessageLink = styled( Link )`
	color: ${ colors.$color_black };
	text-decoration: underline;
	&:hover, &:focus {
		color: ${ colors.$color_blue };
	}
`;

const Block = styled.div`
	display: inline-block;
	margin-right: 32px;
	vertical-align: middle;
`;

const RenewalImage = styled.img`
	width: 150px;
`;

/**
 * Renewal notification message
 *
 * To help the customer understand what happened to licenses.
 *
 * @param {object} props Properties of the component.
 * @returns {ReactElement} Subscription component.
 * @constructor
 */
class RenewalNotification extends React.Component {
	/**
	 * Initializes the renewal notification.
	 */
	constructor()  {
		super();
		this.state = {
			hide: false,
		};
		this.onCrossClick = this.onCrossClick.bind( this );
	}

	componentDidMount() {
		this.props.loadData();
	}

	/**
	 * Called on cross click.
	 *
	 * @param {event} event The event clicking the close cross.
	 *
	 * @returns {void}
	 */
	onCrossClick( event ) {
		event.preventDefault();
		this.setState( { hide: true } );
	}

	/**
	 * Renders the message.
	 *
	 * @returns { ReactElement|null} Returns a message container including the renewalNotification or null.
	 */
	render() {
		if ( this.props.upcomingRenewals.length < 1 ) {
			return null;
		}

		// The first item in the array is the first upcoming renewal (sorted in the container).
		const earliestRenewal = this.props.upcomingRenewals[ 0 ];

		if ( this.state.hide || ! earliestRenewal ) {
			return null;
		}

		const expiryDate = <FormattedDate
			value={ earliestRenewal.nextPayment }
			year="numeric"
			month="long"
			day="2-digit"
		/>;

		return (
			<MessageContainer>
				<CloseButtonTopRight
					onClick={ this.onCrossClick }
					aria-label={ this.props.intl.formatMessage( messages.close ) }
				/>
				<Block>
					<RenewalImage src={ renewalNotification } />
				</Block>
				<Block>
					<h2>
						<FormattedMessage { ...messages.header } />
					</h2>
					<p>
						<FormattedMessage
							{ ...messages.description }
							values={ { expiryDate: expiryDate } }
						/>
						<br />
						<MessageLink
							to={ earliestRenewal.renewalUrl }
							linkTarget="_blank"
						>
							<FormattedMessage { ...messages.linkMessage } />
							<NewTabMessage />
						</MessageLink>
					</p>
				</Block>
			</MessageContainer>
		);
	}
}

RenewalNotification.propTypes = {
	intl: intlShape.isRequired,
	upcomingRenewals: PropTypes.array,
};

RenewalNotification.defaultProps = {
	upcomingRenewals: [],
};

export default injectIntl( RenewalNotification );
