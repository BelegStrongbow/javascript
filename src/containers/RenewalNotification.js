import { connect } from "react-redux";
import { getAllSubscriptions } from "../actions/subscriptions";
import RenewalNotification from "../components/RenewalNotification";
import without from "lodash/without";

/* eslint-disable require-jsdoc */
export const mapStateToProps = ( state ) => {
	const allIds = state.entities.subscriptions.allIds;

	let upcomingRenewals = allIds.map( ( subscriptionId ) => {
		const subscription = state.entities.subscriptions.byId[ subscriptionId ];

		const nextPayment = subscription.nextPayment ? new Date( subscription.nextPayment ) : null;
		const monthFromNow = new Date();
		monthFromNow.setMonth( monthFromNow.getMonth() + 1 );

		const isUpcomingRenewal = subscription.status === "active" &&
			! subscription.endDate &&
			subscription.renewalUrl &&
			nextPayment && nextPayment < monthFromNow;

		if ( ! isUpcomingRenewal ) {
			return null;
		}

		return {
			id: subscription.id,
			name: subscription.name,
			hasNextPayment: subscription.nextPayment !== null,
			nextPayment,
			endDate: subscription.endDate ? new Date( subscription.endDate ) : null,
			status: subscription.status,
			renewalUrl: subscription.renewalUrl,
		};
	} );

	// Removing nulls from the array.
	upcomingRenewals = without( upcomingRenewals, null );

	// Sorting upcoming renewals by nextPayment date.
	upcomingRenewals = upcomingRenewals.sort( ( a, b ) => {
		return new Date( a.nextPayment ) - new Date( b.nextPayment );
	} );

	return {
		upcomingRenewals,
	};
};

export const mapDispatchToProps = ( dispatch ) => {
	return {
		loadData: () => {
			dispatch( getAllSubscriptions() );
		},
	};
};

const RenewalNotificationContainer = connect(
	mapStateToProps,
	mapDispatchToProps
)( RenewalNotification );

export default RenewalNotificationContainer;
