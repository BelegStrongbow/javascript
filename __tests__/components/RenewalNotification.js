import React from 'react';
import { createComponentWithIntl } from "../../utils";

import RenewalNotification from '../../src/components/RenewalNotification';

test('The RenewalNotification component matches the snapshot', () => {
	const component = createComponentWithIntl(
		<RenewalNotification
			upcomingRenewals={ [
				{
					id: "4",
					icon: "https://yoast-mercury.s3.amazonaws.com/uploads/2013/02/Yoast_Icon_Large_RGB.png",
					name: "SEO Premium for WordPress",
					used: 3,
					limit: 20,
					hasNextPayment: true,
					nextPayment: "12-01-2019",
					endDate: null,
					hasEndDate: false,
					billingAmount: 12512,
					billingCurrency: "USD",
					status: "active",
					renewalUrl: "https://my.yoast.test"
				},
			] }
			loadData={ () => {} }
		/>
	);

	let tree = component.toJSON();
	expect(tree).toMatchSnapshot();
});

