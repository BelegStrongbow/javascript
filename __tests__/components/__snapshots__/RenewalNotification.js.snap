// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The RenewalNotification component matches the snapshot 1`] = `
.c3 {
  height: 48px;
  margin: 0;
  padding: 0 16px;
  border: 0;
  background-color: #64a60a;
  color: #fff;
  box-shadow: inset 0 -4px 0 rgba(0,0,0,0.2);
  -webkit-filter: drop-shadow(0 2px 4px rgb(0,0,0,0.2));
  filter: drop-shadow(0 2px 4px rgb(0,0,0,0.2));
  border-radius: 4px;
  font: 700 14px/24px "Open Sans",sans-serif;
  cursor: pointer;
  vertical-align: top;
  -webkit-transition: background-color 150ms ease-out;
  transition: background-color 150ms ease-out;
}

.c3:hover,
.c3:focus {
  box-shadow: 0 4px 10px 0 rgba(0,0,0,0.2),inset 0 0 0 100px rgba( 0,0,0,0.1 );
}

.c3:active {
  -webkit-transform: translateY( 1px );
  -ms-transform: translateY( 1px );
  transform: translateY( 1px );
  box-shadow: none;
}

.c2 {
  background-repeat: no-repeat;
  background-image: url( times.svg );
  background-position: 8px 50%;
  background-size: 24px;
  padding-left: 36px;
}

.c1 {
  background-color: transparent;
  background-position: 50%;
  padding-left: 0px;
  height: 32px;
  width: 32px;
  border-bottom: none;
  box-shadow: none;
  display: inline-block;
}

.c1:hover,
.c1:focus {
  box-shadow: none;
  background-color: transparent;
  opacity: .5;
}

.c7 {
  -webkit-transform: translateY(1em);
  -ms-transform: translateY(1em);
  transform: translateY(1em);
}

.c0 {
  background-color: #fff;
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.2);
  padding: 8px 32px;
  margin-bottom: 32px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 button {
  position: absolute;
  right: 16px;
  top: 16px;
}

.c6 {
  color: #000;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c6:hover,
.c6:focus {
  color: #0066cd;
}

.c4 {
  display: inline-block;
  margin-right: 32px;
  vertical-align: middle;
}

.c5 {
  width: 150px;
}

<div
  className="c0"
>
  <button
    aria-label="close"
    className="c1 c2 c3"
    onClick={[Function]}
    type="button"
  />
  <div
    className="c4"
  >
    <img
      className="c5"
      src="renewalNotification.svg"
    />
  </div>
  <div
    className="c4"
  >
    <h2>
      <span>
        One or more plugins are about to expire!
      </span>
    </h2>
    <p>
      <span>
        When they do, you will no longer receive (security) updates or support! The first one expires on 
        <span>
          October 09, 2018
        </span>
        .
      </span>
      <br />
      <a
        className="c6"
        href="https://my.yoast.test"
        rel="noopener noreferrer"
        target="_blank"
      >
        <span>
          Renew your subscription(s) on this page.
        </span>
        <span
          className="screen-reader-text c7"
        >
          <span>
            (Opens in a new browser tab)
          </span>
        </span>
      </a>
    </p>
  </div>
</div>
`;
